sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -117,6 +118,57 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     }
   }
 
+  override def visitCompoundOrSingleStatement(
+      ctx: CompoundOrSingleStatementContext): CompoundBody = {
+    Option(ctx.singleCompoundStatement()).map { s =>
+      visit(s).asInstanceOf[CompoundBody]
+    }.getOrElse {
+      val logicalPlan = visitSingleStatement(ctx.singleStatement())
+      CompoundBody(List(SparkStatementWithPlan(
+        parsedPlan = logicalPlan,
+        sourceStart = ctx.start.getStartIndex,
+        sourceEnd = ctx.stop.getStopIndex + 1)))
+    }
+  }
+
+  override def visitSingleCompoundStatement(ctx: SingleCompoundStatementContext): CompoundBody = {
+    visit(ctx.beginEndCompoundBlock()).asInstanceOf[CompoundBody]
+  }
+
+  private def visitCompoundBodyImpl(ctx: CompoundBodyContext): CompoundBody = {
+    val buff = ListBuffer[CompoundPlanStatement]()
+    for (i <- 0 until ctx.getChildCount) {
+      val child = ctx.getChild(i)
+      val visitedChild = visit(ctx.getChild(i))
```

Comment:
when will it be null?

Reply:
https://github.com/apache/spark/pull/46665#discussion_r1625787748

Basically, the only case I have found this to be null is when child is instance of `TerminalNodeImpl`. However, for the sake of resiliency and unexpected stuff happening here in future, I added this `case null if child.isInstanceOf[TerminalNodeImpl] =>` case below.

You can check my thoughts in the discussion I pasted above.